### DISCUSSION TITLE
[RESEARCH] Other Mapping Methods Instead of SLAM

### DISCUSSION THREAD

**Discussion by pablo-collantes** [Comet-Robotics/URC#9] [Autonomous]
For the basics of our autonomous, we need to simply use the depth camera to flag obstacles and still understand where they are once we look away. 

SLAM could do this through its localization method by taking encoder odometry and other sensor data like the IMU or camera data. 

So we want to look into other methods that could also help create this localization and mapping without having to integrate all of SLAM since most of its features are pretty redundant for us.

**Comment by kevinmsmith214** [comment]
[ROS example of "HandHeld Mapping" with the realsense D435](https://wiki.ros.org/rtabmap_ros/TutorialsNoetic/HandHeldMapping) There is a constraint for using localization when the map isn't large enough

**Comment by kevinmsmith214** [comment]
[Autonomous Navigation using Turtlebot that works with the rtabmap_ros from above example](https://wiki.ros.org/rtabmap_ros/Tutorials/MappingAndNavigationOnTurtlebot)

**Comment by kevinmsmith214** [comment]
I got in touch with a PHD student at UTD that is specializing in autonomous navigation and mapping. I told him what we data we had (GPS, RGB-D, possibly an IMU) and told him the kind of environment that we would be navigating. This was his response.

Thanks, Kevin, for the information. I get the scenario now. Here are some points to consider. 

1. If the GPS unit is turned on and left alone for a considerable amount of time before the start of the challenge time, it should provide a good global position estimate to you. If the GPS unit also houses a magnetometer, then orientation estimate would be available. This along with the IMU should provide you the (x, y, theta)  or the (lat, lon, theta) estimate of the robot's position from its start position.
 
2. Assuming  "1"  works well, you can initialize a grid of desired resolution, size and generate a global plan from point A to point B using search algorithms like A-star or Dijkstra.  Then you can employ a local planner to avoid local obstacles using the D435 module.

3. Let us say, "1" doesn't work out as intended. You can still run a VSLAM like RTABMap or the latest version of ORBSLAM . Even though the terrain would be similar, the ground still provides contrasting features. The key is the camera orientation. If you orient the camera at 45 - 60 degree facing forward, then you should have enough of the ground to have good visual features. While also scanning for any obstacles in front. ( you still have to employ a local planner in addition to SLAM module ) Be sure to have a good stabilizer for the D435 module and also consider only the features that are within 3-4 meters of D435 module since it has only a short depth range.   Many people have used Intel's T265 tracking module in their[ applications](https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fardupilot.org%2Fdev%2Fdocs%2Fros-vio-tracking-camera.html&data=05%7C02%7CKevin.Smith%40UTDallas.edu%7C2a660582ed324f20f5e108dcd4c1fd54%7C8d281d1d9c4d4bf7b16e032d15de9f6c%7C0%7C0%7C638619176892690918%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C0%7C%7C%7C&sdata=H6VJgZpUBG7RlqJ9lc6RtYtYjBI7rCSiie2a%2BGwseZY%3D&reserved=0). Some implementation details which are used before . You can implement a similar version with D435 module.

4. You may explore [octomaps](https://nam02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FOctoMap%2Foctomap&data=05%7C02%7CKevin.Smith%40UTDallas.edu%7C2a660582ed324f20f5e108dcd4c1fd54%7C8d281d1d9c4d4bf7b16e032d15de9f6c%7C0%7C0%7C638619176892702992%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C0%7C%7C%7C&sdata=NhLzbELkmhQdX72oXvn%2FyPHmY4Yd3RVN7wKbnptNWSQ%3D&reserved=0). People use them for 3D mapping. 

5. I'm not much aware of 3D planners for ground robots. One idea is to project the 3D map into 2D and use 2D global and local planners in ROS. This method has some limitations though. 

So, my bet is on the GPS since it would provide the best estimate of the position and yaw angle when in open space.  If not, you can run VSLAM, but with some careful considerations of camera placement. Feel free to drop by if needed. Good luck!

Regards,
Sai Haneesh, Allu
